mantidproject/mantid
# Add options dialog to Reflectometry GUI [PART 1]

**Description of work.**
This work is a combination of several PRs (which have all been reviewed and merged into this branch) to add an Options dialog and related settings to the ISIS reflectometry interface.

**To test:**
Complete a code review. Full manual testing instructions:
- Check that the dialog is displayed when requested
- Check that **warn discard changes** is displayed when checked and expected (e.g. editing and closing a batch, loading a new batch on top of existing changes, etc) and the negative case (e.g. warning should not be displayed if a batch is first saved, then closed/overwritten). The following should now all be correct:

> Works well in these scenarios:

>- Warning appears when loading a batch after table has changed, and respects the setting.
>- Warning appears when starting autoprocess, and respects the setting.
>- Warning does not appear after pausing and restarting autoprocessing even if the table has >changed, because the table does not get overwritten unless the search criteria have changed. The >warning does appear if the search criteria have changed.
>- Warning appears after changing settings on Event/Experiment/Instrument tabs.
>
>However, seems to be incorrect in these scenarios:
>
>- Warning does not appear after editing the Save tab.
>- Warning when starting autoprocess does not respect the setting.
>- Batch is incorrectly marked as unsaved after processing a row (when state has changed but no content in the table has actually changed)
>- Warning incorrectly appears when I `Restore Defaults` in Experiment and Instrument tabs if nothing has changed. I don't think it's necessary to have a warning here so we can just remove this.

- Check that **warn process all/partial** is displayed when checked and expected (e.g. no rows selected, all groups with content selected, all rows selected, etc). The following should all work:

> A few problems showed up in functional testing. I'm working with the INTER_NR_test.json table from the ISIS sample data.

>- The warning correctly appears when partially processing a group but does not respect the setting when I turn it off.
>- If I select all of the rows within a group, I get a spurious warning that the whole table will be processed - if I turn it off, the whole table does indeed get processed, when it should not - this is a regression. (I also noticed that if I re-load the table, the progress bar does not reset, but this is an existing issue so not related to this PR.)
>- If I select a group (but not its contained rows) and process, then I get a warning that the group will not be fully processed, when actually it will.
>- If I select a group, including its contained rows, and then process, then I get a crash.

-  Check that the **table precision** is displayed as expected (e.g. default value if unchecked, user-specified value if checked, default value if unchecked again, etc).
- Check that all tests pass.

Fixes #26533 and fixes #22051 

*This does not require release notes* because **they will be added in a subsequent PR**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
